feat: compute_consume_specific_shared_reservation

gryczj opened the pull request:
## Description

Fixes #<ISSUE-NUMBER>

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [x] I have followed guidelines from [CONTRIBUTING.MD](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md) and [Samples Style Guide](https://googlecloudplatform.github.io/samples-style-guide/)
- [x] **Tests** pass:   `npm test` (see [Testing](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#run-the-tests-for-a-single-sample))
- [x] **Lint** pass:   `npm run lint` (see [Style](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#style))
- [ ] These samples need a new **API enabled** in testing projects to pass (let us know which ones)
- [ ] These samples need a new/updated **env vars** in testing projects set to pass (let us know which ones)
- [x] This pull request is from a branch created directly off of `GoogleCloudPlatform/nodejs-docs-samples`. Not a fork.
- [ ] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CODEOWNERS) with the codeowners for this sample
- [ ] This sample adds a new sample directory, and I created [GitHub Actions workflow](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/CONTRIBUTING.md#adding-new-samples) for this sample
- [ ] This sample adds a new **Product API**, and I updated the [Blunderbuss issue/PR auto-assigner](https://github.com/GoogleCloudPlatform/nodejs-docs-samples/blob/main/.github/blunderbuss.yml) with the codeowners for this sample
- [x] Please **merge** this PR for me once it is approved

Commit:
feat: compute_consume_specific_shared_reservation

gryczj commented on compute/test/createComputeHyperdisk.test.js:
```diff
@@ -30,24 +29,13 @@ describe('Create compute hyperdisk', async () => {
-  before(async () => {
-    projectId = await disksClient.getProjectId();
+  after(async () => {
     // Cleanup resources
```
In this PR I also moved cleanup methods to after() to prevent keeping not used resources in our project after running the tests.

m-strzelczyk commented on compute/reservations/createInstanceToConsumeSharedReservation.js:
```diff
@@ -0,0 +1,118 @@
+   * TODO(developer): Update these variables before running the sample.
+   */
+  // The ID of the project where instance will be consumed and created.
+  const baseProjectId = 'base-project-id';
```
Can we rename to a more obvious names? Like `reservationOwnerProjectID` and `reservationConsumerProjectID` for example?

gryczj replied:
@m-strzelczyk sure, fixed

iennae committed:
Merge branch 'main' into compute_consume_specific_shared_reservation

iennae committed:
Merge branch 'main' into compute_consume_specific_shared_reservation

iennae committed:
Merge branch 'main' into compute_consume_specific_shared_reservation

BigBlackWolf commented:
Hi @iennae, could you please take a look once again on this PR?

cc: @rsamborski

iennae committed:
Merge branch 'main' into compute_consume_specific_shared_reservation

iennae committed:
chore(compute): update comments -- remove obvious comments and add additional comments

Obvious comments distract from the sample. Add clarifying comments that explain why and where users might customize.

iennae commented:
thanks @BigBlackWolf for the ping. Can you check the changes I made to the one file to clarify the comments? I think there is a lot of specific choices embedded in the samples that could use a little more detail and then some obvious comments that distract from the code. For example the instantiates aren't giving more context to the sample consumers. Could you make updates to the other files to match that style? 

In general, I know there are other samples like this, but it really feels import with the number of choices that folks have with compute to better understand how to manage that complexity rather than hide it and end up with copied patterns over time with no understanding of the "why". 

Thank you!

iennae committed:
fix(compute): remove the spaces I introduced

iennae reviewed (commented):
Aside from the comments needing improvement, I have some additional concerns. Refactoring to eliminate duplicate code and create helper functions would enhance readability and maintainability. These new helper functions could accept parameters like disk size, source image, and network settings, simplifying updates to this collection of samples and streamlining the samples across different attributes. Currently, the handcrafted samples require extensive reviews and introduce duplicate code.


I will seek a second opinion from the team tomorrow and update this PR accordingly. While I lean towards a conditional approval, I believe we can enhance consistency and style.


Thank you for your patience.

BigBlackWolf commented:
I appreciate your review @iennae and thanks for adjusting comments, imho they became much more informative!
I agree, that it's worth to correct other samples, however it looks like out of the scope of this PR. As I see, the intention was to create **compute_consume_specific_shared_reservation**.
Don't you mind introducing your suggestions in a separate PR?

iennae commented:
Heya, the problem is the other issues I raised and this PR has made me realize there is a lot of repetition in these set of samples whoch is going to make them harder to maintain. Do we have an assigned engineering team that will be ok maintaining these?

BigBlackWolf commented:
Unfortunately, I am not aware of those. 
Do you think changes, that help reduce repetition should be introduced in this PR?

iennae commented:
I think refactoring the new sample that is getting added since it has the same duplicate functions could happen in this PR. Do you know if there is an engineering team that will take on maintenance? My primary concern is the folks who have to maintain these. To be clear, I'm nit saying all the existing samples need to get updated but the new one should be refactoring to make this duplication code maintainable.

m-strzelczyk committed:
Merge branch 'main' into compute_consume_specific_shared_reservation

glasnt commented:
CLA check override note: previous executions worked https://github.com/GoogleCloudPlatform/nodejs-docs-samples/pull/3912/checks?check_run_id=34225819577

glasnt committed:
Merge branch 'main' into compute_consume_specific_shared_reservation
